torch/lib/THCUNN/SoftMaxCommon.cuh

```diff
@@ -0,0 +1,454 @@
+
+        Epilogue<T, AccumT> epilogue(max_input, sum);
+        for (uint32_t d = threadIdx.x; d < dim_size; d += blockDim.x)
+          output[data_offset + d * dim_stride] = epilogue(input[data_offset + d * dim_stride]);
+      } else {
+        T max_input = THCNumerics<T>::min();
+        for (uint32_t d = 0; d < dim_size; d++) {
+          const T value = input[data_offset + d * dim_stride];
+          max_input = THCNumerics<T>::ge(max_input, value) ? max_input : value;
+        }
+
+        AccumT sum = 0;
+        for (uint32_t d = 0; d < dim_size; d++)
+          sum += THCNumerics<T>::exp(input[data_offset + d * dim_stride] - max_input);
+
+        Epilogue<T, AccumT> epilogue(max_input, sum);
+        for (uint32_t d = 0; d < dim_size; d++)
+          output[data_offset + d * dim_stride] = epilogue(input[data_offset + d * dim_stride]);
+      }
+    }
+  }
+}
+
+template <typename T, typename AccumT, template<typename, typename> class Epilogue>
+__global__ void cunn_SpatialSoftMaxBackward(
+    T *gradInput, T *output, T *gradOutput,
+    uint32_t outer_size, uint32_t dim_size, uint32_t inner_size)
+{
+  SharedMem<AccumT> smem;
+  const uint32_t outer_stride = inner_size * dim_size;
+  const uint32_t dim_stride = inner_size;
+
+  for (uint32_t outer_index = blockIdx.x; outer_index < outer_size; outer_index += gridDim.x) {
+    const uint32_t outer_offset = outer_index * outer_stride;
+    for (uint32_t inner_index = blockIdx.y * blockDim.y + threadIdx.y; inner_index < inner_size; inner_index += blockDim.y * gridDim.y) {
+      const uint32_t data_offset = outer_offset + inner_index;
+      // See the comment in forward kernel
+      if (blockDim.x > 1) {
+        AccumT sum = 0;
+        for (uint32_t d = threadIdx.x; d < dim_size; d += blockDim.x)
+          sum += gradOutput[data_offset + d * dim_stride];
+        sum = spatialBlockReduceX<AccumT, Add>(smem.getPointer(), sum);
+
+        Epilogue<T, AccumT> epilogue(sum);
+        for (uint32_t d = threadIdx.x; d < dim_size; d += blockDim.x) {
+          gradInput[data_offset + d * dim_stride] =
+            epilogue(gradOutput[data_offset + d * dim_stride],
+                    output[data_offset + d * dim_stride]);
+        }
+      } else {
+        AccumT sum = 0;
+        for (uint32_t d = 0; d < dim_size; d++)
+          sum += gradOutput[data_offset + d * dim_stride];
+
+        Epilogue<T, AccumT> epilogue(sum);
+        for (uint32_t d = 0; d < dim_size; d++) {
+          gradInput[data_offset + d * dim_stride] =
+            epilogue(gradOutput[data_offset + d * dim_stride],
+                    output[data_offset + d * dim_stride]);
+        }
+      }
+    }
+  }
+}
+
+////////////////////////////////////////////////////////////////////////////////
+// Regular kernel (fast when dim_size is large; requires inner_size == 1)
+////////////////////////////////////////////////////////////////////////////////
+
+
+template <typename T, typename AccumT>
+struct MaxFloat
```

Comment:
Bikeshed comment: are these supposed to be for Floats only or any type T?

Reply:
It's an old piece of code. I only wrote/refactored the spatial case and left the old last-dim only algorithm mostly untouched. In here it actually makes sense because it uses this `f` function (now I'm not sure if we want to keep this behavior for double, it might have been missed when double was enabled in THCUNN)

Reply:
Yeah I don't recognize the `fmaxType` command - is it specializing to `fmax` or `fmaxf` as appropriate?. It seems like you would want to ScalarConvert `v` to `AccumT` and then do the op.

Reply:
Ah this is actually correct. `fmax` should have an overload for double, so no need to change this